File: companion/bolumler/b08/ornek-01/cozum.py
import argparse
import math
from numbers import Real
from pathlib import Path
import platform

import numpy as np
import pandas as pd
import scipy
from scipy import stats


def hazirla(veri):
    if list(veri.columns) != ["n", "ortalama", "s"] or len(veri) != 1:
        raise ValueError("Tek satirda n, ortalama, s ozeti gerekli; bu dosya ham veri degildir.")
    if veri.isna().any().any():
        raise ValueError("Eksik ozet degeri kabul edilmez.")
    for sutun in veri.columns:
        if not pd.api.types.is_numeric_dtype(veri[sutun]) or pd.api.types.is_bool_dtype(veri[sutun]):
            raise ValueError("Sayisal ozet degerleri gerekli.")
    if not np.isfinite(veri.to_numpy(dtype=float)).all():
        raise ValueError("Sonlu ozet degerleri gerekli.")
    temiz = veri.reset_index(drop=True).copy(deep=True)
    hacim = float(temiz.loc[0, "n"])
    if hacim < 2 or not hacim.is_integer():
        raise ValueError("n en az 2 olan tam sayi olmali.")
    if temiz.loc[0, "s"] <= 0:
        raise ValueError("Bu pilot pozitif orneklem standart sapmasi gerektirir.")
    return temiz


def aralik_hesapla(veri, duzey):
    temiz = hazirla(veri)
    if not isinstance(duzey, Real) or isinstance(duzey, (bool, np.bool_)) or not math.isfinite(duzey) or not 0 < duzey < 1:
        raise ValueError("Guven duzeyi 0 ile 1 arasinda olmali; 95 yerine 0.95 kullanin.")
    hacim = int(temiz.loc[0, "n"])
    merkez = float(temiz.loc[0, "ortalama"])
    standart_hata = float(temiz.loc[0, "s"]) / math.sqrt(hacim)
    kritik = float(stats.t.ppf((1 + duzey) / 2, df=hacim - 1))
    hata_payi = kritik * standart_hata
    sonuc = {"guven_duzeyi": duzey, "alpha": 1 - duzey, "kritik_t": kritik,
             "hata_payi": hata_payi, "alt_sinir": merkez - hata_payi,
             "ust_sinir": merkez + hata_payi, "genislik": 2 * hata_payi}
    if not all(math.isfinite(deger) for deger in sonuc.values()):
        raise ValueError("Aralik sayisal olarak hesaplanamadi.")
    return sonuc


def hesapla(veri):
    temiz = hazirla(veri)
    hacim = int(temiz.loc[0, "n"])
    satirlar = [("ozet", "hacim", hacim), ("ozet", "ortalama", temiz.loc[0, "ortalama"]),
                ("ozet", "orneklem_sd", temiz.loc[0, "s"]),
                ("ozet", "standart_hata", temiz.loc[0, "s"] / math.sqrt(hacim)),
                ("ozet", "serbestlik", hacim - 1)]
    for etiket, duzey in [("GA95", 0.95), ("GA99", 0.99)]:
        satirlar.extend((etiket, olcu, deger) for olcu, deger in aralik_hesapla(temiz, duzey).items())
    return pd.DataFrame(satirlar, columns=["degisken", "olcu", "deger"])


def kontrol_et(sonuc, beklenen):
    if list(beklenen.columns) != ["degisken", "olcu", "deger"]:
        raise ValueError("Kontrol sutunlari uyusmuyor.")
    if not sonuc[["degisken", "olcu"]].equals(beklenen[["degisken", "olcu"]]):
        raise ValueError("Kontrol etiketleri veya sirasi uyusmuyor.")
    hedefler = pd.to_numeric(beklenen["deger"], errors="raise")
    if not np.isfinite(hedefler).all() or not np.isfinite(sonuc["deger"]).all():
        raise ValueError("Sonlu kontrol degerleri gerekli.")
    for gercek, hedef in zip(sonuc["deger"], hedefler):
        if not math.isclose(gercek, hedef, rel_tol=1e-9, abs_tol=1e-9):
            raise ValueError(f"Kontrol uyusmazligi: {gercek} != {hedef}")
    print(f"DOGRULANDI: {len(sonuc)} kontrol degeri eslesiyor.")


def grafik_kaydet(veri):
    import matplotlib
    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    temiz = hazirla(veri)
    merkez = float(temiz.loc[0, "ortalama"])
    fig, eksen = plt.subplots(figsize=(9, 3.5))
    for konum, duzey, renk in [(1, 0.95, "#1F4E79"), (0, 0.99, "#087F5B")]:
        ozet = aralik_hesapla(temiz, duzey)
        eksen.errorbar(merkez, konum, xerr=ozet["hata_payi"], fmt="o", capsize=7,
                       color=renk, linewidth=2, markersize=7)
        eksen.annotate(f'{ozet["alt_sinir"]:.4f}', (ozet["alt_sinir"], konum),
                      xytext=(0, 12), textcoords="offset points", ha="center")
        eksen.annotate(f'{ozet["ust_sinir"]:.4f}', (ozet["ust_sinir"], konum),
                      xytext=(0, 12), textcoords="offset points", ha="center")
    eksen.axvline(merkez, color="#AAAAAA", linestyle=":")
    eksen.set(yticks=[0, 1], yticklabels=["%99 guven", "%95 guven"], ylim=(-0.5, 1.6),
              xlabel="Evren ortalamasi icin aralik (puan)",
              title=f'n={int(temiz.loc[0, "n"])}; ortalama={merkez:g}; s={temiz.loc[0, "s"]:g}')
    eksen.margins(x=0.15)
    fig.tight_layout()
    hedef = Path("ciktilar/python")
    hedef.mkdir(parents=True, exist_ok=True)
    fig.savefig(hedef / "guven-araliklari.png", dpi=150)
    plt.close(fig)
    print("Grafik: ciktilar/python/guven-araliklari.png")


def main():
    parser = argparse.ArgumentParser(description="B08: ozet istatistiklerden t guven araligi")
    parser.add_argument("--check", action="store_true")
    parser.add_argument("--grafik", action="store_true")
    secenek = parser.parse_args()
    veri = pd.read_csv("veri.csv")
    sonuc = hesapla(veri)
    print(sonuc.to_string(index=False))
    if secenek.check:
        kontrol_et(sonuc, pd.read_csv("beklenen-sonuclar.csv"))
    if secenek.grafik:
        grafik_kaydet(veri)
    print(f"Python {platform.python_version()}, NumPy {np.__version__}, pandas {pd.__version__}, SciPy {scipy.__version__}")


if __name__ == "__main__":
    main()

File: companion/bolumler/b08/ornek-01/cozum.R
hazirla <- function(veri) {
  if (!identical(names(veri), c("n", "ortalama", "s")) || nrow(veri) != 1) {
    stop("Tek satirda n, ortalama, s ozeti gerekli; bu ham veri degildir.")
  }
  if (!all(vapply(veri, is.numeric, logical(1))) || anyNA(veri) ||
      any(!is.finite(as.matrix(veri)))) stop("Sonlu ve eksiksiz sayisal ozet gerekli.")
  if (veri$n < 2 || veri$n != floor(veri$n)) stop("n en az 2 olan tam sayi olmali.")
  if (veri$s <= 0) stop("Bu pilot pozitif orneklem standart sapmasi gerektirir.")
  veri
}

aralik_hesapla <- function(veri, duzey) {
  temiz <- hazirla(veri)
  if (!is.numeric(duzey) || length(duzey) != 1 || !is.finite(duzey) || duzey <= 0 || duzey >= 1) {
    stop("Guven duzeyi 0 ile 1 arasinda olmali; 95 yerine 0.95 kullanin.")
  }
  standart_hata <- temiz$s / sqrt(temiz$n)
  kritik <- qt((1 + duzey) / 2, df = temiz$n - 1)
  hata_payi <- kritik * standart_hata
  sonuc <- c(guven_duzeyi = duzey, alpha = 1 - duzey, kritik_t = kritik,
              hata_payi = hata_payi, alt_sinir = temiz$ortalama - hata_payi,
              ust_sinir = temiz$ortalama + hata_payi, genislik = 2 * hata_payi)
  if (any(!is.finite(sonuc))) stop("Aralik sayisal olarak hesaplanamadi.")
  sonuc
}

hesapla <- function(veri) {
  temiz <- hazirla(veri)
  sonuc <- data.frame(degisken = rep("ozet", 5),
    olcu = c("hacim", "ortalama", "orneklem_sd", "standart_hata", "serbestlik"),
    deger = c(temiz$n, temiz$ortalama, temiz$s, temiz$s / sqrt(temiz$n), temiz$n - 1),
    stringsAsFactors = FALSE)
  for (index in 1:2) {
    duzey <- c(0.95, 0.99)[index]
    aralik <- aralik_hesapla(temiz, duzey)
    sonuc <- rbind(sonuc, data.frame(degisken = rep(c("GA95", "GA99")[index], 7),
                  olcu = names(aralik), deger = as.numeric(aralik)))
  }
  sonuc
}

kontrol_et <- function(sonuc, beklenen) {
  if (!identical(names(beklenen), c("degisken", "olcu", "deger")) ||
      !identical(sonuc$degisken, beklenen$degisken) ||
      !identical(sonuc$olcu, beklenen$olcu)) stop("Kontrol etiketleri uyusmuyor.")
  if (!is.numeric(beklenen$deger) || anyNA(beklenen$deger) ||
      any(!is.finite(beklenen$deger)) || any(!is.finite(sonuc$deger))) stop("Sonlu kontrol degerleri gerekli.")
  tolerans <- pmax(1e-9, 1e-9 * pmax(abs(sonuc$deger), abs(beklenen$deger)))
  if (any(abs(sonuc$deger - beklenen$deger) > tolerans)) stop("Kontrol uyusmazligi.")
  cat("DOGRULANDI:", nrow(sonuc), "kontrol degeri eslesiyor.\n")
}

grafik_kaydet <- function(veri) {
  temiz <- hazirla(veri)
  aralik95 <- aralik_hesapla(temiz, 0.95)
  aralik99 <- aralik_hesapla(temiz, 0.99)
  genislik <- aralik99["genislik"]
  dir.create("ciktilar/r", recursive = TRUE, showWarnings = FALSE)
  png("ciktilar/r/guven-araliklari.png", width = 1350, height = 525, res = 150)
  on.exit(dev.off())
  par(mar = c(4, 6, 3, 1))
  plot(NA, xlim = c(aralik99["alt_sinir"], aralik99["ust_sinir"]) + c(-0.15, 0.15) * genislik,
       ylim = c(-0.5, 1.6), yaxt = "n", xlab = "Evren ortalamasi icin aralik (puan)", ylab = "",
       main = paste("n=", temiz$n, "; ortalama=", temiz$ortalama, "; s=", temiz$s))
  axis(2, at = 0:1, labels = c("%99 guven", "%95 guven"), las = 1)
  abline(v = temiz$ortalama, col = "grey", lty = 3)
  for (index in 1:2) {
    aralik <- list(aralik95, aralik99)[[index]]
    konum <- 2 - index
    renk <- c("#1F4E79", "#087F5B")[index]
    segments(aralik["alt_sinir"], konum, aralik["ust_sinir"], konum, col = renk, lwd = 2)
    points(temiz$ortalama, konum, pch = 19, col = renk)
    for (uc in aralik[c("alt_sinir", "ust_sinir")]) {
      segments(uc, konum - 0.08, uc, konum + 0.08, col = renk)
      text(uc, konum + 0.23, sprintf("%.4f", uc))
    }
  }
  cat("Grafik: ciktilar/r/guven-araliklari.png\n")
}

secenek <- commandArgs(trailingOnly = TRUE)
if (any(!secenek %in% c("--check", "--grafik"))) stop("Bilinmeyen secenek.")
veri <- read.csv("veri.csv", check.names = FALSE)
sonuc <- hesapla(veri)
print(sonuc, row.names = FALSE)
if ("--check" %in% secenek) kontrol_et(sonuc, read.csv("beklenen-sonuclar.csv", stringsAsFactors = FALSE))
if ("--grafik" %in% secenek) grafik_kaydet(veri)
sessionInfo()